source/elements/oneMKL/source/domains/spblas/operations/spmv.rst

```diff
@@ -0,0 +1,378 @@
+           oneapi::mkl::sparse::matrix_handle_t       A_handle,
+           oneapi::mkl::sparse::dense_vector_handle_t x_handle,
+           void*                                      beta,
+           oneapi::mkl::sparse::dense_vector_handle_t y_handle,
+           oneapi::mkl::sparse::spmv_alg              alg,
+           oneapi::mkl::sparse::spmv_descr_t          spmv_descr,
+           std::size_t                                &temp_buffer_size,
+       );
+
+       void spmv_optimize(
+           sycl::queue                                &queue,
+           oneapi::mkl::transpose                     opA,
+           void*                                      alpha,
+           oneapi::mkl::sparse::matrix_view           A_view,
+           oneapi::mkl::sparse::matrix_handle_t       A_handle,
+           oneapi::mkl::sparse::dense_vector_handle_t x_handle,
+           void*                                      beta,
+           oneapi::mkl::sparse::dense_vector_handle_t y_handle,
+           oneapi::mkl::sparse::spmv_alg              alg,
+           oneapi::mkl::sparse::spmv_descr_t          spmv_descr,
+           sycl::buffer<std::uint8_t, 1>              workspace,
+       );
+
+       sycl::event spmv_optimize(
+           sycl::queue                                &queue,
+           oneapi::mkl::transpose                     opA,
+           void*                                      alpha,
+           oneapi::mkl::sparse::matrix_view           A_view,
+           oneapi::mkl::sparse::matrix_handle_t       A_handle,
+           oneapi::mkl::sparse::dense_vector_handle_t x_handle,
+           void*                                      beta,
+           oneapi::mkl::sparse::dense_vector_handle_t y_handle,
+           oneapi::mkl::sparse::spmv_alg              alg,
+           oneapi::mkl::sparse::spmv_descr_t          spmv_descr,
+           void*                                      workspace,
+           const std::vector<sycl::event>             &dependencies = {},
+       );
+
+       sycl::event spmv(
+           sycl::queue                                &queue,
+           oneapi::mkl::transpose                     opA,
+           void*                                      alpha,
+           oneapi::mkl::sparse::matrix_view           A_view,
+           oneapi::mkl::sparse::matrix_handle_t       A_handle,
+           oneapi::mkl::sparse::dense_vector_handle_t x_handle,
+           void*                                      beta,
+           oneapi::mkl::sparse::dense_vector_handle_t y_handle,
+           oneapi::mkl::sparse::spmv_alg              alg,
+           oneapi::mkl::sparse::spmv_descr_t          spmv_descr,
+           const std::vector<sycl::event>             &dependencies = {},
+       );
+
+   }
+
+.. container:: section
+
+   .. rubric:: Notes
+
+   - ``spmv_buffer_size`` and ``spmv_optimize`` must be called at least once before ``spmv``
+     with the same arguments. ``spmv`` can then be called multiple times.
```

Comment:
How about allowing calling `spmv` directly with `spmv_descr == nullptr`, in which case the operation must be attempted without any extra memory being allocated. The operation may fail and throw an exception if it can't proceed without that temp workspace.

Is this something we can/should allow?

Reply:
Possibly, we can allow this for `spmv`/`spmm`, but disallow it for `spsv`/`spsm`.

Reply:
With our choice to keep `optimize_*` APIs compulsory with a separate algorithm parameter that is a placeholder for doing nothing/minimal things in `optimize` stage, I think we can don't need to consider `spmv_descr == nullptr` case. You can resolve this thread.